beacon-chain/core/state/state_transition.go

```diff
@@ -26,81 +30,96 @@ func NewStateTransition(
 	justifiedSlot := st.LastJustifiedSlot()
 	finalizedSlot := st.LastFinalizedSlot()
 	timeSinceFinality := block.SlotNumber() - newState.LastFinalizedSlot()
-	recentBlockHashes := st.RecentBlockHashes()
 	newState.SetValidators(v.CopyValidators(newState.Validators()))
 
-	if st.LastStateRecalculationSlot() < params.BeaconConfig().CycleLength {
-		lastStateRecalculationSlotCycleBack = 0
-	} else {
-		lastStateRecalculationSlotCycleBack = st.LastStateRecalculationSlot() - params.BeaconConfig().CycleLength
-	}
-
-	// walk through all the slots from LastStateRecalculationSlot - cycleLength to
-	// LastStateRecalculationSlot - 1.
-	for i := uint64(0); i < params.BeaconConfig().CycleLength; i++ {
-		var blockVoteBalance uint64
-
-		slot := lastStateRecalculationSlotCycleBack + i
-		blockHash := recentBlockHashes[i]
-
-		blockVoteBalance, validators := incentives.TallyVoteBalances(
-			blockHash,
-			blockVoteCache,
-			newState.Validators(),
-			v.ActiveValidatorIndices(newState.Validators()),
-			v.TotalActiveValidatorDeposit(newState.Validators()),
-			timeSinceFinality,
-		)
-
-		newState.SetValidators(validators)
-
-		justifiedSlot, finalizedSlot, justifiedStreak = FinalizeAndJustifySlots(
-			slot,
-			justifiedSlot,
-			finalizedSlot,
-			justifiedStreak,
-			blockVoteBalance,
-			v.TotalActiveValidatorDeposit(st.Validators()),
-		)
-	}
-
-	crossLinks, err := crossLinkCalculations(
-		newState,
-		st.PendingAttestations(),
-		block.SlotNumber(),
-	)
+	newState.ClearAttestations(st.LastStateRecalculationSlot())
+	// Derive the new set of recent block hashes.
+	recentBlockHashes, err := st.CalculateNewBlockHashes(block, parentSlot)
 	if err != nil {
 		return nil, err
 	}
+	newState.SetRecentBlockHashes(recentBlockHashes)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update recent block hashes: %v", err)
+	}
 
-	newState.SetCrossLinks(crossLinks)
+	newRandao := createRandaoMix(block.RandaoReveal(), st.RandaoMix())
+	newState.SetRandaoMix(newRandao[:])
 
-	newState.SetLastJustifiedSlot(justifiedSlot)
-	newState.SetLastFinalizedSlot(finalizedSlot)
-	newState.SetJustifiedStreak(justifiedStreak)
-	newState.SetLastStateRecalculationSlot(newState.LastStateRecalculationSlot() + params.BeaconConfig().CycleLength)
+	// The changes below are only applied if this is a cycle transition.
+	if block.SlotNumber()%params.BeaconConfig().CycleLength == 0 {
```

Comment:
Why did this logic change from before? A cycle transition can also occur if `block.slot % cyleLength == 1` if the previous slot was skipped, and this needs to also handle the corner case where multiple cycles worth of slots are skipped.